cmd/lotus-shed/fip-0036.go

```diff
@@ -0,0 +1,565 @@
+				votesIncludingMsigs[signerId] = APPROVE
+				clientApproveBytes = big.Add(clientApproveBytes, clientBytes)
+			} else {
+				rejectionBalance = types.BigAdd(rejectionBalance, accountActor.Balance)
+				votesIncludingMsigs[signerId] = Reject
+				clientRejectBytes = big.Add(clientRejectBytes, clientBytes)
+			}
+
+			//process msigs
+			// TODO: Oh god, oh god, there's a possibility that a multisig has voted in both directions
+			// and we'll pick a winner non-deterministically as we iterate...
+			// We need to factor in vote time if that happens and pick whoever went first
+			if mss, found := accountsToMultisigs[signerId]; found {
+				for _, ms := range mss { //get all the msig signer has
+					if _, ok := votedMsigs[ms]; ok {
+						// msig has already voted, skip
+						continue
+					}
+					if mpv, found := msigPendingVotes[ms]; found { //other signers of the multisig have voted, yet the threshold has not met
+						if v == APPROVE {
+							if mpv.ApproveCount+1 == mpv.Multisig.Threshold { //met threshold
+								approveBalance = types.BigAdd(approveBalance, mpv.Multisig.Balance)
+								delete(msigPendingVotes, ms) //threshold, can skip later signer votes
+								votedMsigs[ms] = struct{}{}
+								votesIncludingMsigs[ms] = APPROVE
+
+							} else {
+								mpv.ApproveCount++
+								msigPendingVotes[ms] = mpv
+							}
+						} else {
+							if mpv.RejectCount+1 == mpv.Multisig.Threshold { //met threshold
+								rejectionBalance = types.BigAdd(rejectionBalance, mpv.Multisig.Balance)
+								delete(msigPendingVotes, ms) //threshold, can skip later signer votes
+								votedMsigs[ms] = struct{}{}
+								votesIncludingMsigs[ms] = Reject
+
+							} else {
+								mpv.RejectCount++
+								msigPendingVotes[ms] = mpv
+							}
+						}
+					} else { //first vote received from one of the signers of the msig
+						msi, ok := msigActorsInfo[ms]
+						if !ok {
+							return xerrors.Errorf("didn't find msig %s in msig map", ms)
+						}
+
+						if msi.Threshold == 1 { //met threshold with this signer's single vote
+							if v == APPROVE {
+								approveBalance = types.BigAdd(approveBalance, msi.Balance)
+								votesIncludingMsigs[ms] = APPROVE
+
+							} else {
+								rejectionBalance = types.BigAdd(rejectionBalance, msi.Balance)
+								votesIncludingMsigs[ms] = Reject
+							}
+							votedMsigs[ms] = struct{}{}
+						} else { //threshold not met, add to pending vote
+							if v == APPROVE {
+								msigPendingVotes[ms] = msigVote{
+									Multisig:     msi,
+									ApproveCount: 1,
+								}
+							} else {
+								msigPendingVotes[ms] = msigVote{
+									Multisig:    msi,
+									RejectCount: 1,
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// time to process miners based on what we know about votesIncludingMsigs
+		minerVotes := make(map[address.Address]uint64)
+		fmt.Println("counting miner votes")
+		for s, v := range votes {
+			if minerInfos, found := ownerMap[s]; found {
```

Comment:
for s is msig  - this is not checking against whether the music mets the threshold, add a voteMsig found check?

Reply:
oops! that's supposed to be iterating over votesIncludingMsigs -- are you proposing just doing this loop twice, once for votes and once for votedMsigs? I do like that more.

Reply:
im confused - then where is the miner balance when the owner is not msig accounted?

Reply:
oh here https://github.com/filecoin-project/lotus/blob/9928c1ac185be0ecd15f5faec18fc5772e47f904/cmd/lotus-shed/fip-0036.go#L353..